File: archive/r/rust/prime-number.rs
//  Requirement     https://sample-programs.therenegadecoder.com/projects/prime-number/
// Accept a number on command line and print if it is Composite or Prime 
// Works till  39 digits, ...

use std::env::args;
use std::process::exit;
use std::num::ParseIntError;

fn usage() -> ! {
    println!("Usage: please input a non-negative integer");
    exit(0);
}

fn parse_int(s: String) -> Result<i128, ParseIntError> {
    s.trim().parse::<i128>()
}

fn main() {
    // Exit if 1st command-line argument not an integer
    let mut input_value: Result<i128, ParseIntError> = parse_int(
        args().nth(1).unwrap_or_else(|| usage())
    );
    if input_value.is_err() {
        usage();
    }

    // Exit if negative integer
    let input_num: i128 = input_value.unwrap();
    if input_num < 0 {
        usage();
    }

    let mut n = 3 as u128;
    let value = input_num as u128;

    if value < 2 || (value != 2 && value % 2 == 0) {
        println!("Composite");
        exit(0);
    }
    while n * n <= value {
        if value % n == 0 {
            println!("Composite");
            exit(0);
        }
        n = n + 2;
    }
    println!("Prime");
}

File: archive/c/c/prime-number.c
#include <stdio.h>
#include <stdlib.h>

void prime_number(int number)
{
    if (number < 0) {
        printf("Usage: please input a non-negative integer\n");
        return;
    } else if (number == 0 || number == 1) {
        printf("Composite\n");
        return;
    }

    for (int i = 2; i < number; i++) {
        if (number % i == 0) {
            printf("Composite\n");
            return;
        }
    }
    printf("Prime\n");
}

int is_int(char *str)
{
    if (str[0] > '9' || str[0] < '0')
            return (1);
    for (int i = 0; str[i]; i++) {
        if (str[i] > '9' || str[i] < '0')
            return (1);
    }
    return (0);
}

int main(int ac, char **av)
{
    if (ac != 2 || is_int(av[1]) != 0) {
        printf("Usage: please input a non-negative integer\n");
        return (1);
    }
	prime_number(atoi(av[1]));
	return (0);
}
